kvdb-web/src/indexed_db.rs

```diff
@@ -0,0 +1,227 @@
+		.expect_throw("Should be able to open IndexDB");
+
+	try_create_object_stores(&open_request, columns);
+
+	let on_success = Closure::once(move |event: &Event| {
+		// Extract database handle from the event
+		let target = event.target().expect_throw("Event should have a target");
+		let req = target.dyn_ref::<IdbRequest>().expect_throw("Event target is IdbRequest");
+
+		let result = req
+			.result()
+			.expect_throw("IndexedDB.onsuccess should have a valid result");
+		assert!(result.is_instance_of::<IdbDatabase>());
+
+		// errors if the receiving end was dropped before this call
+		let _ = tx.send(IdbDatabase::from(result));
+	});
+	open_request.set_onsuccess(Some(on_success.as_ref().unchecked_ref()));
+	on_success.forget();
+
+	rx.then(|r| future::ready(r.expect("Sender isn't dropped; qed")))
+}
+
+fn store_name(num: u32) -> String {
+	format!("col{}", num)
+}
+
+fn column_to_number(column: Column) -> u32 {
+	column.map(|c| c + 1).unwrap_or_default()
+}
+
+
+// Returns js objects representing store names for each column
+fn store_names_js(columns: u32) -> Array {
+	let column_names = (0..=columns).map(store_name);
+
+	let js_array = Array::new();
+	for name in column_names {
+		js_array.push(&JsValue::from(name));
+	}
+
+	js_array
+}
+
+fn try_create_object_stores(req: &IdbOpenDbRequest, columns: u32) {
+	let on_upgradeneeded = Closure::once(move |event: &Event| {
+		debug!("Upgrading or creating the database");
+		// Extract database handle from the event
+		let target = event.target().expect_throw("Event should have a target");
+		let req = target.dyn_ref::<IdbRequest>().expect_throw("Event target is IdbRequest");
+		let result = req.result().expect_throw("IdbRequest should have a result");
+		let db: &IdbDatabase = result.unchecked_ref();
+
+		for name in (0..=columns).map(store_name) {
+			let res = db.create_object_store(name.as_str());
+			if let Err(err) = res {
+				debug!("error creating object store {}: {:?}", name, err);
+			}
+		}
+	});
+
+	req.set_onupgradeneeded(Some(on_upgradeneeded.as_ref().unchecked_ref()));
+	on_upgradeneeded.forget();
+}
+
+/// Commit a transaction to the IndexedDB.
+pub fn idb_commit_transaction(
+	idb: &IdbDatabase,
+	txn: &DBTransaction,
+	columns: u32,
+) -> impl Future<Output = ()> {
+	let store_names_js = store_names_js(columns);
+
+	// Create a transaction
+	let mode = IdbTransactionMode::Readwrite;
+	let idb_txn = idb.transaction_with_str_sequence_and_mode(&store_names_js, mode)
+		.expect_throw("Failed to create an IndexedDB transaction");
+
+	// Open object stores (columns)
+	let object_stores = (0..=columns).map(|n| {
+		idb_txn.object_store(store_name(n).as_str())
+			.expect("Object stores were created in try_create_object_stores; qed")
+	}).collect::<Vec<_>>();
+
+	for op in &txn.ops {
+		match op {
+			DBOp::Insert { col, key, value } => {
+				let column = column_to_number(*col) as usize;
+
+				// Convert rust bytes to js arrays
+				// TODO (optimize): we could avoid copies
```

Comment:
That'd a bit too dangerous IMO. You don't know what the lifetime of the `Uint8Array` is.